ui/packages/catalog-ui-search/src/main/webapp/js/controllers/ol.layerCollection.controller.js

```diff
@@ -12,197 +12,178 @@
+    return createTile({ ...opts, imageExtent, ...opts.parameters }, ol.source.ImageStatic, ol.layer.Image);
+}
+
+const sources = { OSM, BM, WMS, WMT, AGM, SI }
+
+const createLayer = (type, opts) => {
+    const fn = sources[type];
+    if (fn !== undefined) {
+        return fn(opts);
+    } else {
+        console.warn(`Unknown map layer type ${type}`);
+    }
+}
+
+const Controller = CommonLayerController.extend({
+    initialize: function () {
+        // there is no automatic chaining of initialize.
+        CommonLayerController.prototype.initialize.apply(this, arguments);
+    },
+    makeMap: function (options) {
+        this.collection.forEach((model) => {
+            this.addLayer(model);
+        })
+
+        const view = new ol.View({
+            projection: ol.proj.get(properties.projection),
+            center: ol.proj.transform([0, 0], 'EPSG:4326', properties.projection),
+            zoom: options.zoom
+        });
+
+        const config = {
+            target: options.element,
+            view: view,
+            interactions: ol.interaction.defaults({doubleClickZoom: false})
+        };
+
+        if (options.controls !== undefined) {
+            config.controls = options.controls;
+        }
+
+        this.map = new ol.Map(config);
+        this.isMapCreated = true;
+        return this.map;
+    },
+    onDestroy: function () {
+        if (this.isMapCreated) {
+            this.map.setTarget(null);
+            this.map = null;
         }
-    });
+    },
+    addLayer: async function (model) {
+        const { id, type } = model.toJSON();
+        const opts = _.omit(model.attributes, 'type', 'label', 'index', 'modelCid');
+        opts.show = model.shouldShowLayer();
 
-    return Controller;
+        const layer = await Promise.resolve(createLayer(type, opts));
 
+        if (layer !== undefined) {
+            this.map.addLayer(layer);
+            this.layerForCid[id] = layer;
+            this.reIndexLayers();
+        }
+    },
+    removeLayer: function (model) {
+        const id = model.get('id');
+        const layer = this.layerForCid[id];
+        if (layer !== undefined) {
+            this.map.removeLayer(layer);
+        }
+        delete this.layerForCid[id];
+        this.reIndexLayers();
+    },
+    setAlpha: function (model) {
+        const layer = this.layerForCid[model.id];
+        if (layer !== undefined) {
+            layer.setOpacity(model.get('alpha'));
+        }
+    },
+    setShow: function (model) {
+        const layer = this.layerForCid[model.id];
+        if (layer !== undefined) {
+            layer.setVisible(model.shouldShowLayer());
+        }
+    },
+    reIndexLayers: function () {
+        this.collection.forEach(function (model, index) {
+            const layer = this.layerForCid[model.id];
+            if (layer !== undefined) {
+                layer.setZIndex(-(index + 1));
+            }
+        }, this);
+        user.savePreferences();
```

Comment:
I'm not sure when we should save user preferences. @andrewkfiedler is there a better time to save them?